### PR TITLE
Handle TOTP and yubikey OTP tokens in passwords [WIP]

### PR DESCRIPTION
**Pull request by ryskov**
Now handles both TOTP and Yubikey OTP tokens added to passwords. 

If both types of OTP tokens are configured for user, either one will work.

### CODE REVIEW

**Comment by benyanke** [comment]
On first glance - this looks fantastic. Could make this very useful.

However, I'm floundering a bit to actually get it to work.

With the PR in the readme file, or at the very least, in the comments here (and I can document later), could you provide some brief notes on how to set this up and get it working? I have a yubikey (multiple, actually), so I'd love to test, I'm just not even sure where to start on finding the values needed for the config options - both the global config options (`yubikeyclientid`, `yubikeysecret`), and the user-specific ones (`otpsecret`, `yubikey`).

TLDR: could you add some docs, or at least how-to?

**Comment by ryskov** [comment]
Added OTP configuration guide in the README :)

**Comment by benyanke** [comment]
This looks great - I'm going to have to block out some time to really test this, but it looks good at first glance.

**Comment by woopstar** [comment]
What is the timeframe on these PR's ?

**Comment by benyanke** [comment]
While I don't have quite as much time on my plate as I'd like, this one is next on my list to start testing.

**Comment by benyanke** [comment]
Hey Ryan - spent a bit this weekend digging in, and still having trouble with this one... perhaps it's my still-growing knowledge of some of the more obscure attributes of LDAP.

Do you have any good examples of being able to validate these using a tool like `ldapsearch`? I'm trying to find options for this or similar tools and running by trying to bind as the user simply comes up invalid creds.

Would love your thoughts on this

**Comment by ryskov** [comment]
You are correct, the OTP needs to be appended to the password. I don't know if that is desired, but that fits into our usecase. I think it is a good choice that users should provide both a password and their token to ensure 2fa. 

This is an ldapsearch query that should work with both yubikey and totp tokens (notice the yubikey otp appended to 'mysecret')

`ldapsearch -LLL -H ldap://localhost:3893 -D cn=serviceuser,ou=svcaccts,dc=glauth,dc=com -w mysecretvvjrcfdljltbkeuuurlbjgeetecfurjrecdgglrrrrni -x -bdc=glauth,dc=com \(objectClass=posixaccount\)`

Have you checked out the updated README yet? There should be some guidance and links to setting up the yubikey for OTP

**Comment by benyanke** [comment]
Sorry for the long delay - I'm going to have to reject this one after doing some further looking and thinking. 

The implementation of asking the client to arbitrarily concatenate an additional factor to the known password is quite nonstandard (I have never seen this before and couldn't find an example of it being used anywhere). It also breaks many uses of LDAP. 

It's interesting, to be sure! Just not a great fit for the project here. Thanks again for your contribution, and extremely sorry for the delay! Have had little time for FOSS lately.

**Comment by woopstar** [comment]
Your research is terrible bad then. This is the default behavior of FreeIPA and privacyIDEA to implement 2FA. It is generally used when you want to add lets say 2FA for VPN, in the case where you have OpenVPN and a LDAP backend.

**Comment by fbartels** [comment]
I have also not seen such a combination of password and token in a single field, but for most users in https://www.reddit.com/r/shittyprogramming/comments/90bpsn/charles_schwabs_approach_to_2fa/ this also seems to feel pretty normal. Maybe worth reconsidering?

**Comment by benyanke** [comment]
Well, it seems I may have misjudged the commonality of using a concatenated password and OTP.

I'll reopen this and get it merged.

**Comment by benyanke** [comment]
Hi all - this is merged to dev, master, and released in v1.1.0, with binarirs in the GitHub releases, and images in docker hub.
